# What is pandas
In Python, pandas is a data manipulating library. It is the code word for Panel Data.
To be able to work with the package it need to be loaded before it can be utilized.
The community has agreed to use pd as the alias when importing pandas


In [6]:
import pandas as pd

# How does pandas represent data

pandas represent data in the form of a table with indexed rows and columns.
For instance, we can represent the sales made by 3 sales persons of a tea shop with 3 different columns,
- their names "NAME", 
- the amount of "SALES" made, 
- and finaly categorically "CHECK" if they have closed sales "CLOSED" or not "NOT CLOSED".

# pandas can handle 7 data types(dtype), namely 
- object : Text or mixed numeric and non-numeric values
- int64 : Integer numbers
- float64 : Floating point numbers
- bool : True/False values
- datetime64 : Date and time values
- timedelta[ns] : Differences between two datetimes
- category : Finite list text values

So our "Name" column will be characters/strings(textual data).
The "SALES" column will be integers(Numerical values).
And the "CHECK" column is of an ordinal category(also textual data).

To manually store this data in a table, we create a DataFrame. Using a Python dictionary of list, the dictionary keys will be used as column headers and the values in each list as rows of the DataFram which is in our case correspond to each sales person in the code below.

In [8]:
df = pd.DataFrame({
    'NAME': ['Kwadwo', 'Kodua', 'Britwum'],
    'SALES': [20000, 30000, 40000],
    'CHECK': ['CLOSED', 'CLOSED', 'NOT CLOSED']
})

df

NAME  SALES       CHECK
0   Kwadwo  20000      CLOSED
1    Kodua  30000      CLOSED
2  Britwum  40000  NOT CLOSED

The above illustrates the DataFrame.

# What is a DataFrame

A DataFrame is a 2-dimensional data structure that can store data of different types.
Each column in a DataFrame is a Series

We are only interested in working with the data in the column "SALES".

In [9]:
df['SALES']


0    20000
1    30000
2    40000
Name: SALES, dtype: int64

When selecting a single column of a pandas DataFrame, the result is a pandas Series. Conceptually the column labels serve as the keys in accessing the columns as illustrate above. df is the name of our DataFrame.

We can also create a Series from scratch as well


In [10]:
Sales = pd.Series([20000, 30000, 40000], name='SALES')

Sales

0    20000
1    30000
2    40000
Name: SALES, dtype: int64

A pandas Series has no column labels, as it is just a single column of a DataFrame. As Series does does have row labels(indices starting from 0)

# We can perform functions and operations on       DataFrames

For example we can apply the method max() to the "SALES" column of the DataFrame "df" to know the maximum sales made.

In [11]:
df['SALES'].max()

40000

Or to the Series "Sales"

In [12]:
Sales.max()

40000

We can do several things with a DataFrame of Series.
pandas provides a lot of functionalities.
As methons are functions, always use parenthesis().

# We can perform basic statistics of the numerical data of the table.

In [13]:
df.describe()

SALES
count      3.0
mean   30000.0
std    10000.0
min    20000.0
25%    25000.0
50%    30000.0
75%    35000.0
max    40000.0

The describe() method provides a quick overview of the numerical data in a DataFrame. As the "NAME" and "CHECK" columns are textual data, these are by default not taken into account by the describe() method.

Many pandas operations return a DataFrame of a Series. The describe() method is an example of a pandas operation returning a pandas Series.

Apart from the representation, also the data manipulations and calculations you would do in spreadsheet software are supported by pandas.


# Check this out

In [15]:
SalesPersonOne = ['Kwadwo', 20000, 'CLOSED']
SalesPersonTwo = ['Kodua', 30000, 'CLOSED']
SalesPersonThree = ['Britwum', 40000, 'NOT CLOSED']

df2 = pd.DataFrame([SalesPersonOne, SalesPersonTwo, SalesPersonThree],
                  columns = ['NAMES', 'SALES', 'CHECK'])

df2

NAMES  SALES       CHECK
0   Kwadwo  20000      CLOSED
1    Kodua  30000      CLOSED
2  Britwum  40000  NOT CLOSED

# How to read and write tabular data

We can analyse data available as a CSV file.
Now lets load a CSV file of the payment records of permit by Transport Unions in KMA(Kumasi Metropolitan Assembly) "Cash_is_queen" as presented by the Department of Transport.

pandas provides the read_csv() function to read data stored as a csv file into a pandas DataFrame.
pandas support many different file formats of data sources out of the box(csv, excel, sql, parquet,...) each of them with the prefix read_*.
Always type in the whole path to the file you wish to work on.
Make sure to always have a check on the data after reading in the data.
When displaying a DataFrame, the first and last 5 rows will be shown by default:

In [37]:
cash = pd.read_csv("C:\\DataScience\\Cash_is_queen.csv") 

cash

SUB-METRO                               UNION        SERVICE DEMAND YEAR  \
0      Bantama  Commercial Drivers Transport Union           Taxi        2019   
1      Bantama           Concerned Transport Union           Taxi        2019   
2      Bantama           Concerned Transport Union           Taxi        2019   
3      Bantama           Concerned Transport Union         Trotro        2019   
4      Bantama                              PROTOA         Trotro        2019   
...        ...                                 ...            ...         ...   
1000     Subin                               GPRTU          Cargo        2019   
1001     Subin           Concerned Transport Union  Long Distance        2019   
1002  Nhyiaeso                               GPRTU           Taxi        2019   
1003     Subin                               GPRTU  Long Distance        2019   
1004     Subin                              PROTOA  Long Distance        2019   

      DEBIT  CREDIT  
0    -400.0     NaN  
1    -400.0   200.0  
2    -200.0   200.0  
3    -200.0   200.0  
4    -400.0     NaN  
...     ...     ...  
1000 -400.0   200.0  
1001 -400.0   200.0  
1002 -200.0   200.0  
1003 -200.0   200.0  
1004 -400.0   200.0  

[1005 rows x 6 columns]

We can view the first 8 rows of the pandas DataFrame.

In [38]:
cash.head(8)

SUB-METRO                               UNION SERVICE DEMAND YEAR  DEBIT  \
0   Bantama  Commercial Drivers Transport Union    Taxi        2019 -400.0   
1   Bantama           Concerned Transport Union    Taxi        2019 -400.0   
2   Bantama           Concerned Transport Union    Taxi        2019 -200.0   
3   Bantama           Concerned Transport Union  Trotro        2019 -200.0   
4   Bantama                              PROTOA  Trotro        2019 -400.0   
5   Bantama                               GPRTU    Taxi        2019 -200.0   
6   Bantama  Commercial Drivers Transport Union    Taxi        2019 -200.0   
7     Subin           Concerned Transport Union    Taxi        2019 -600.0   

   CREDIT  
0     NaN  
1   200.0  
2   200.0  
3   200.0  
4     NaN  
5   200.0  
6     NaN  
7     NaN

To see the first N rows of a DataFrame, use the head() method with the required number of rows(in this case 8) as argument.

Interested in the last N rows instead? pandas also provides a tail() method.

In [39]:
cash.tail(8)

SUB-METRO                        UNION        SERVICE DEMAND YEAR  DEBIT  \
997      Subin                        GPRTU  Long Distance        2019 -200.0   
998      Subin     Redeemer Transport Union  Long Distance        2019 -200.0   
999      Subin  Cooperative Transport Union  Long Distance        2019 -200.0   
1000     Subin                        GPRTU          Cargo        2019 -400.0   
1001     Subin    Concerned Transport Union  Long Distance        2019 -400.0   
1002  Nhyiaeso                        GPRTU           Taxi        2019 -200.0   
1003     Subin                        GPRTU  Long Distance        2019 -200.0   
1004     Subin                       PROTOA  Long Distance        2019 -400.0   

      CREDIT  
997    200.0  
998    200.0  
999    200.0  
1000   200.0  
1001   200.0  
1002   200.0  
1003   200.0  
1004   200.0

A check on how pandas interpreted each of the column data types can be done by requesting the pandas 'dtypes' attribute

In [40]:
cash.dtypes

SUB-METRO       object
UNION           object
SERVICE         object
DEMAND YEAR     object
DEBIT          float64
CREDIT         float64
dtype: object

For each of the columns, the used data type is enlisted. The data types in our DataFrame are strings(object) and floats(float64)

When aking for the dtypes, no brackets are used. dtypes is an attribute of a DataFrame and Series. Attributes of DataFrame aof Series do not need brackets. Atttibutes represent a characteristic of a DataFrame/Series, whereas methods(which requires brackets) do something with the DataFrame/Series.

We can save a spreadsheet version of our cash data.

In [49]:
cash.to_excel('C:\\DataScience\\cash.xlsx', sheet_name='permitRecords', index=False)

Whereas read_* functions are used to read data to pandas, the to_* methods are used to store data. The to_excel() method stores the data as an excel file. In the example here, the sheet_name is named "permitRecords" instead of the default "Sheet1". By setting index=False the row index label are not saved in the spreadsheet.

Now we use the read_excel() method to read the excel file we have created

In [59]:
cash2 = pd.read_excel('cash.xlsx', sheet_name='permitRecords')

cash2

SUB-METRO                               UNION        SERVICE DEMAND YEAR  \
0      Bantama  Commercial Drivers Transport Union           Taxi        2019   
1      Bantama           Concerned Transport Union           Taxi        2019   
2      Bantama           Concerned Transport Union           Taxi        2019   
3      Bantama           Concerned Transport Union         Trotro        2019   
4      Bantama                              PROTOA         Trotro        2019   
...        ...                                 ...            ...         ...   
1000     Subin                               GPRTU          Cargo        2019   
1001     Subin           Concerned Transport Union  Long Distance        2019   
1002  Nhyiaeso                               GPRTU           Taxi        2019   
1003     Subin                               GPRTU  Long Distance        2019   
1004     Subin                              PROTOA  Long Distance        2019   

      DEBIT  CREDIT  
0      -400     NaN  
1      -400   200.0  
2      -200   200.0  
3      -200   200.0  
4      -400     NaN  
...     ...     ...  
1000   -400   200.0  
1001   -400   200.0  
1002   -200   200.0  
1003   -200   200.0  
1004   -400   200.0  

[1005 rows x 6 columns]

We can view some technical summary of a DataFrame using the mathod info().

In [60]:
cash2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005 entries, 0 to 1004
Data columns (total 6 columns):
SUB-METRO      1004 non-null object
UNION          1004 non-null object
SERVICE        1004 non-null object
DEMAND YEAR    1005 non-null object
DEBIT          1005 non-null int64
CREDIT         187 non-null float64
dtypes: float64(1), int64(1), object(4)
memory usage: 47.2+ KB


Now let's explain the output in more detail:
- It is indeed a DataFrame
- There are 1005 entries, i.e. 1005 rows.
- The table has 6 columns. Two columns have a value for each of the rows(all 1005 values are non-null). Some columns do have missing values thus less than 1005 non-null values.
- The columns "SUB-METRO", "UNION", "SERVICE" and "DEMAND YEAR"     consist of textual data(strings, aka object). The "DEBIT" column consists of whole numbers(aka integer) and "CREDIT" column consists of real numbers(aka float)
- The kind of data in the different columns are summarized by the dtypes listing.
- The approximate amount of RAM used to hold the DataFrame is provided as well.

# How to select a subset of a DataFrame.

How to select specfic column from our DataFram.

Let's select the "DEBIT" column of the cash2 DataFram

In [62]:
debit = cash2['DEBIT']

debit.head()

0   -400
1   -400
2   -200
3   -200
4   -400
Name: DEBIT, dtype: int64

To select a single column, use square brackets[] with the column of interest as illustrated above.

Each column in a DataFrame is a Series. As a single column is selected, the returned object is a pandas DataFrame. We can verify this by checking the type of the output.

In [63]:
type(debit)

pandas.core.series.Series

And have a look at the shape of the output

In [64]:
cash2['DEBIT'].shape 

(1005,)

In [65]:
cash2.shape

(1005, 6)

DataFrame.shape is an attribute(remember not to use paranthesis) of a pandas Series and DataFrame containing the number of rows and columns:(nrows, ncolumns). A pandas Series is 1-dimensional and only the number of rows is returned.

We can select the "DEBIT" and "CREDIT" columns


In [68]:
debit_credit = cash2[['DEBIT', 'CREDIT']]

debit_credit

DEBIT  CREDIT
0      -400     NaN
1      -400   200.0
2      -200   200.0
3      -200   200.0
4      -400     NaN
...     ...     ...
1000   -400   200.0
1001   -400   200.0
1002   -200   200.0
1003   -200   200.0
1004   -400   200.0

[1005 rows x 2 columns]

To select multiple columns, use a list of column names within the selection brackets[].

The inner square brackets define a Python list with column names, whereas the outer brackets are used to select the data from a pandas DataFrame as seen in the previous example

The returned data type is a pandas DataFrame:


In [69]:
type(cash2[['DEBIT', 'CREDIT']])

pandas.core.frame.DataFrame

In [70]:
debit_credit.shape

(1005, 2)

The selection returned a DataFrame with 1005 rows and 2 columns. Remember a DataFrame is 2-dimensional with both a row and column dimension.

# How to filter specific rows from a DataFrame

Now let's select Unions with debit exceeding -200
Note that debit is a negative value.


In [72]:
below_neg200 = cash2[cash2['DEBIT'] < -200]

below_neg200.head()

SUB-METRO                               UNION SERVICE DEMAND YEAR  DEBIT  \
0    Bantama  Commercial Drivers Transport Union    Taxi        2019   -400   
1    Bantama           Concerned Transport Union    Taxi        2019   -400   
4    Bantama                              PROTOA  Trotro        2019   -400   
7      Subin           Concerned Transport Union    Taxi        2019   -600   
10   Bantama           Concerned Transport Union    Taxi        2019   -400   

    CREDIT  
0      NaN  
1    200.0  
4      NaN  
7      NaN  
10     NaN

To select rows based on a conditional expression, use a condition inside the selection []
The condition inside the selection brackets cash2['DEBIT'] < -200 checks for which rows in the "DEBIT" column has a value exceeding -200

In [73]:
cash2['DEBIT'] < -200

0        True
1        True
2       False
3       False
4        True
        ...  
1000     True
1001     True
1002    False
1003    False
1004     True
Name: DEBIT, Length: 1005, dtype: bool

The output of the conditional expressions is actually a pandas Series of boolean values(either True of False) with the same number of rows as the original DataFrame. Such a Series of boolean values can be used to filter the DataFrame by putting it in between the selection brackets[]. Only rows for which the value is True will be selected.

We know from before that the original cash2 DataFrame consists of 1005 row. Let's have a look at the amount of rows which satisfy the condition by checking the shape attribute of the resulting DataFrame below_neg200: 


In [74]:
below_neg200.shape

(639, 6)

In [79]:
taxis_and_cargos = cash2[cash2['SERVICE'].isin(['Taxi','Cargo'])]

taxis_and_cargos

SUB-METRO                               UNION SERVICE DEMAND YEAR  DEBIT  \
0      Bantama  Commercial Drivers Transport Union    Taxi        2019   -400   
1      Bantama           Concerned Transport Union    Taxi        2019   -400   
2      Bantama           Concerned Transport Union    Taxi        2019   -200   
5      Bantama                               GPRTU    Taxi        2019   -200   
6      Bantama  Commercial Drivers Transport Union    Taxi        2019   -200   
...        ...                                 ...     ...         ...    ...   
988      Subin                               GPRTU    Taxi        2019   -200   
989      Subin                                 ***    Taxi        2019   -600   
990    Bantama                               GPRTU    Taxi        2019   -200   
1000     Subin                               GPRTU   Cargo        2019   -400   
1002  Nhyiaeso                               GPRTU    Taxi        2019   -200   

      CREDIT  
0        NaN  
1      200.0  
2      200.0  
5      200.0  
6        NaN  
...      ...  
988      NaN  
989      NaN  
990    200.0  
1000   200.0  
1002   200.0  

[131 rows x 6 columns]

Similar to the conditional expression, the isin() conditional function returns a True for each row the values are in the provided list. To filter the rows based on such a function, use the conditional function inside the selection brackets []. In this case, the condition inside the selection brackets cash2['SERVICE'].isin(['Taxi', 'Cargo']) checks for which rows the "SERVICE" column is either Taxi or Cargo.

The above is equivalent to filtering by rows for which the class is either Taxi or Cargo and combining the two statements with an | (or) operator



In [84]:
taxis_and_cargos2 = cash2[(cash2['SERVICE']=='Taxi') | (cash2['SERVICE']=='Cargo')]

taxis_and_cargos2

SUB-METRO                               UNION SERVICE DEMAND YEAR  DEBIT  \
0      Bantama  Commercial Drivers Transport Union    Taxi        2019   -400   
1      Bantama           Concerned Transport Union    Taxi        2019   -400   
2      Bantama           Concerned Transport Union    Taxi        2019   -200   
5      Bantama                               GPRTU    Taxi        2019   -200   
6      Bantama  Commercial Drivers Transport Union    Taxi        2019   -200   
...        ...                                 ...     ...         ...    ...   
988      Subin                               GPRTU    Taxi        2019   -200   
989      Subin                                 ***    Taxi        2019   -600   
990    Bantama                               GPRTU    Taxi        2019   -200   
1000     Subin                               GPRTU   Cargo        2019   -400   
1002  Nhyiaeso                               GPRTU    Taxi        2019   -200   

      CREDIT  
0        NaN  
1      200.0  
2      200.0  
5      200.0  
6        NaN  
...      ...  
988      NaN  
989      NaN  
990    200.0  
1000   200.0  
1002   200.0  

[131 rows x 6 columns]

The above is same as using the isin() method.
When combining multiple conditional statements, each condition must be surrounded by parenthesis (). Moreover, you can not use or/and but instead use '|' as (or) and '&' as (and).

Now we want to work with cash2 data for which the "CREDIT" has a value

In [85]:
credit_notna = cash2[cash2['CREDIT'].notna()]

credit_notna

SUB-METRO                      UNION        SERVICE DEMAND YEAR  DEBIT  \
1      Bantama  Concerned Transport Union           Taxi        2019   -400   
2      Bantama  Concerned Transport Union           Taxi        2019   -200   
3      Bantama  Concerned Transport Union         Trotro        2019   -200   
5      Bantama                      GPRTU           Taxi        2019   -200   
9      Bantama                      GPRTU           Taxi        2019   -200   
...        ...                        ...            ...         ...    ...   
1000     Subin                      GPRTU          Cargo        2019   -400   
1001     Subin  Concerned Transport Union  Long Distance        2019   -400   
1002  Nhyiaeso                      GPRTU           Taxi        2019   -200   
1003     Subin                      GPRTU  Long Distance        2019   -200   
1004     Subin                     PROTOA  Long Distance        2019   -400   

      CREDIT  
1      200.0  
2      200.0  
3      200.0  
5      200.0  
9      200.0  
...      ...  
1000   200.0  
1001   200.0  
1002   200.0  
1003   200.0  
1004   200.0  

[187 rows x 6 columns]

The notna() conditional function returns a True for each row that the values are not a Null/NaN value. As such, this can be combined with the selection brackets[] to filter the data table.

We can find out what actually changed by verifying using the 'shape' attribute

In [86]:
credit_notna.shape

(187, 6)

# How to select specific rows and columns from a DataFrame

We are noe interested in the unions and their corresponding services that have made some payments.

In [92]:
credited_unions = cash2.loc[cash2['CREDIT'] > 0, ['UNION','SERVICE']]

credited_unions

UNION        SERVICE
1     Concerned Transport Union           Taxi
2     Concerned Transport Union           Taxi
3     Concerned Transport Union         Trotro
5                         GPRTU           Taxi
9                         GPRTU           Taxi
...                         ...            ...
1000                      GPRTU          Cargo
1001  Concerned Transport Union  Long Distance
1002                      GPRTU           Taxi
1003                      GPRTU  Long Distance
1004                     PROTOA  Long Distance

[187 rows x 2 columns]

In this case, a subset of both rows and columns is made in one go and just using selection brackets[] is not sufficient anymore. The loc/iloc operators are required in front of the selection bracket[]. When using loc/iloc, the part before the comma is the rows you want, and the part after the comma is the columns you want to select.

When using the column names, row labels or a condition expression, use the 'loc' operator in front of the selection brackets[]. For both the part before and after the comma, you can use a single label, a list of labels, a slice of labels, a conditional expression or a colon. Using a colon specifies you want to select all rows or columns.

Now let's select rows 10 till 25 and columns 3 to 5

In [93]:
cash2.iloc[9:25, 2:5]

SERVICE DEMAND YEAR  DEBIT
9            Taxi        2019   -200
10           Taxi        2019   -400
11           Taxi        2019   -200
12  Long Distance        2019   -200
13  Long Distance        2019   -600
14           Taxi        2019   -400
15           Taxi        2019   -600
16           Taxi        2019   -600
17           Taxi        2019   -600
18         Trotro        2019   -200
19         Trotro        2019   -400
20  Long Distance        2019   -200
21  Long Distance        2019   -400
22  Long Distance        2019   -200
23  Long Distance        2019   -400
24  Long Distance        2019   -400

Again, a subset of both rows and columns is made in one go and just using selection bracket[] is not sufficient anymore. When specifically interested in certain rows and/or columns based on their position in the table, use the iloc operator in front of the selection brackets[].

When selecting specific rows and/or columns with loc or iloc, new values can be assigned to the selected data. For example, to assign the name anonymous to the first 3 elements of the second column:

In [94]:
cash2.iloc[0:3, 2 ] = 'anonymous'

cash2.head()

SUB-METRO                               UNION    SERVICE DEMAND YEAR  DEBIT  \
0   Bantama  Commercial Drivers Transport Union  anonymous        2019   -400   
1   Bantama           Concerned Transport Union  anonymous        2019   -400   
2   Bantama           Concerned Transport Union  anonymous        2019   -200   
3   Bantama           Concerned Transport Union     Trotro        2019   -200   
4   Bantama                              PROTOA     Trotro        2019   -400   

   CREDIT  
0     NaN  
1   200.0  
2   200.0  
3   200.0  
4     NaN

Now let's save to file

In [95]:
cash2.to_excel('C:\\DataScience\\cash.xlsx', sheet_name='permitRecords', index=False)

# Notice:

- When selecting subsets of data, square bracket[] are used.
- Inside these brackets, you can use a single column/row label, a list of column/row labels, a slice of labels, a conditional expression or a colon
- Select specific rows and/or columns using loc when using the row and column names.
- Select specific rows and/or columns using iloc when using the position in the table
- You can assign new values to a selection based on loc/iloc